pkg/auth/handler/sso/sso.go

```diff
@@ -0,0 +1,174 @@
+package sso
+
+import (
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/provider/oauth"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/provider/password"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/sso"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/userprofile"
+	"github.com/skygeario/skygear-server/pkg/auth/response"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authtoken"
+	"github.com/skygeario/skygear-server/pkg/core/auth/role"
+	"github.com/skygeario/skygear-server/pkg/server/skydb"
+	"github.com/skygeario/skygear-server/pkg/server/skyerr"
+)
+
+type respHandler struct {
+	RoleStore            role.Store
+	TokenStore           authtoken.Store
+	AuthInfoStore        authinfo.Store
+	OAuthAuthProvider    oauth.Provider
+	PasswordAuthProvider password.Provider
+}
+
+func (h respHandler) loginActionResp(oauthAuthInfo sso.AuthInfo) (resp interface{}, err error) {
+	// action => login
+	var info authinfo.AuthInfo
+	err = h.handleLogin(&info, oauthAuthInfo)
+	if err != nil {
+		return
+	}
+
+	// Create auth token
+	var token authtoken.Token
+	token, err = h.TokenStore.NewToken(info.ID)
+	if err != nil {
+		panic(err)
+	}
+	if err = h.TokenStore.Put(&token); err != nil {
+		panic(err)
+	}
+
+	// TODO: convert oauthAuthInfo.UserProfile to userprofile.UserProfile
+	var userProfile userprofile.UserProfile
+	resp = response.NewAuthResponse(info, userProfile, token.AccessToken)
+
+	// Populate the activity time to user
+	now := timeNow()
+	info.LastSeenAt = &now
+	if err = h.AuthInfoStore.UpdateAuth(&info); err != nil {
+		err = skyerr.MakeError(err)
+		return
+	}
+
+	return
+}
+
+func (h respHandler) linkActionResp(oauthAuthInfo sso.AuthInfo) (resp interface{}, err error) {
+	// action => link
+	// check if provider user is already linked
+	_, err = h.OAuthAuthProvider.GetPrincipalByProviderUserID(oauthAuthInfo.ProviderName, oauthAuthInfo.ProviderUserID)
+	if err == nil {
+		err = skyerr.NewError(skyerr.InvalidArgument, "user linked to the provider already")
+		return resp, err
+	}
+
+	if err != skydb.ErrUserNotFound {
+		// some other error
+		return resp, err
+	}
+
+	// check if user is already linked
+	userID := oauthAuthInfo.State.UserID // skygear userID
+	_, err = h.OAuthAuthProvider.GetPrincipalByUserID(userID)
+	if err == nil {
+		err = skyerr.NewError(skyerr.InvalidArgument, "provider account already linked with existing user")
```

Comment:
We should check the oauth principal with both user id and provider name, one user should be enable to connect with both google and facebook account. But they cannot connect with 2 different google accounts.